### PR TITLE
HDDS-3371. Begin Cleanup of old write-path of key in OM

### PR DESCRIPTION
**Pull request by GeorgeJahad**
## What changes were proposed in this pull request?

When the OM-HA code was added, the tests were not completely updated.  Many of them still call the OzoneManagerProtocol "write-path" methods even though those methods have been superseded by the HA methods.

This PR is a small beginning to the fix.  I didn't want to do too much before getting some feedback, so I just removed a single api call,  OzoneManager.deleteKey(), and fixed the corresponding tests, by changing them to invoke the ozoneManagerClient version.

If there is a less hacky way of doing it, without using the client, please let me know.  I couldn't find one.  I tried creating an alternative, but it looked like a half-baked client, so I just used the real one.

I timed all the tests I changed and didn't notice any running longer.

### Notes on TestOmMetrics.testKeyOps() test

The only test that was difficult to fix was the testKeyOps() test [here](https://github.com/apache/ozone/blob/79a9d39da7f33e71bc00183e280105562354cca4/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmMetrics.java#L243).

This test verifies that keyOps generate the correct number of metrics.  I tried to write the new code so the numbers didn't change.  But there is actually a flaw in the original test that causes it to undercount a few, (*NumKeyLookupFails*, and *NumKeyDeleteFails*).  So you'll see those fixed.

The test depends on mocks of the KeyManager. But now that the OM no longer supports deleteKey(), those mocks won't work.  So I had to replace most of them with actual working code; (because I couldn't find a way to mock the actual write path.)

The test is divided into two halves.  The first half verifies normal metrics; the second half verifies failure metrics.  I removed the mocks in the first half; as mentioned above, those metrics are now generated by real code.  The mocks in the second half are there to force failures, (by generating exceptions).  I modified those to work with the new code.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3371

## How was this patch tested?

All effected tests were updated.

### CODE REVIEW

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmMetrics.java [hunk 1]
```diff
@@ -276,33 +279,45 @@ public void testKeyOps() throws IOException {
...
+
+    // inject exception to test for Failure Metrics on the write path
+    OMMetadataManager metadataManager = (OMMetadataManager)
+        HddsWhiteboxTestUtils.getInternalState(ozoneManager, "metadataManager");
+    OMMetadataManager mockMm = Mockito.spy(metadataManager);
+    Table<String, OmBucketInfo> bucketTable = (Table<String, OmBucketInfo>)
+        HddsWhiteboxTestUtils.getInternalState(metadataManager, "bucketTable");
+    Table<String, OmBucketInfo> mockBTable = Mockito.spy(bucketTable);
+    Mockito.doThrow(exception).when(mockBTable).isExist(any());
+    Mockito.doReturn(mockBTable).when(mockMm).getBucketTable();
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmMetrics.java:316]: Spotbugs reported this about this line so I added an exemption:
```
M D RV: Return value of OMMetadataManager.getBucketTable() ignored, but method has no side effect  At TestOmMetrics.java:[line 316]"
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmMetrics.java [hunk 2]
```diff
@@ -276,33 +279,45 @@ public void testKeyOps() throws IOException {
...
 
-
-    ozoneManager.openKey(keyArgs);
-    ozoneManager.commitKey(keyArgs, 0);
-    ozoneManager.openKey(keyArgs);
-    ozoneManager.commitKey(keyArgs, 0);
-    ozoneManager.openKey(keyArgs);
-    ozoneManager.commitKey(keyArgs, 0);
-    ozoneManager.deleteKey(keyArgs);
+    keyArgs = createKeyArgs();
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmMetrics.java:284]: In the previous version the mocks would allow the counters would increment, even though the same key was being written every time.  To allow this to work now, I had to create new keyArgs, each time a new key was committed.

**Comment by GeorgeJahad** [comment]
@bharatviswa504 This is just the start of what you requested, but I didn't want to go too far without you confirming that my approach was right.  So please take a look when you get a chance.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 3]
```diff
@@ -2682,27 +2682,8 @@ public void renameKey(OmKeyArgs args, String toKeyName) throws IOException {
...
-      AUDIT.logWriteSuccess(buildAuditMessageForSuccess(OMAction.DELETE_KEY,
-          auditMap));
-      metrics.decNumKeys();
-    } catch (Exception ex) {
-      metrics.incNumKeyDeleteFails();
-      AUDIT.logWriteFailure(buildAuditMessageForFailure(OMAction.DELETE_KEY,
-          auditMap, ex));
-      throw ex;
-    }
+    throw new UnsupportedOperationException("OzoneManager does not require " +
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2685]: Just a question, instead of doing this way, is there a way we can remove these methods.

My question is coming from the context, as if any new methods added to OzoneManagerProtocol, still should add this UnsupportedOperationException?

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2685]: If that is too much change, we can explore that in a new Jira. Lets get the work going in, as I see it is causing lot of confusion to OM developers

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2685]: And also in similar mannner are you also planning to delete KeyManagerImpl delete method also? (I think we can completely remove that method)

**Comment by bharatviswa504** [comment]
> @bharatviswa504 This is just the start of what you requested, but I didn't want to go too far without you confirming that my approach was right. So please take a look when you get a chance.

Thank You @GeorgeJahad for the work, over all approach LGTM. Few questions I have commented them inline

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 3 above)
Comment (cxorm) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2685]: I'm +1 for removing these methods.

IMHO there are two ways to do it,

1. Change the `OzoneManagerProtocol`. 
> It is a **huge** patch, and we would encounter many issues like backward-compatibility and API compatibility.

2. Deprecating `OzoneManager.java`.
> In this case, we should update those part of using `OzoneManager.java`. (Still a `huge` patch.) 

These ways are too complicated to archive in a clear manner. 
I ever used to the same throw-Exception manner and I think it doesn't solve the problem fully.

Feel free to correct me if you have any idea.

Comment (GeorgeJahad) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2685]: > Just a question, instead of doing this way, is there a way we can remove these methods.

I haven't looked at that much yet, because I saw that deleteKeys(), renameKeys(), setQuota() and revokeS3Secret() all use the UnsupportedOperationException approach, so I assumed that is what I was supposed to do.  I just copied the code from deleteKeys()

> My question is coming from the context, as if any new write requests
>  added to OzoneManagerProtocol, still we should add this
>  UnsupportedOperationException?

I think so, unfortunately, until we redefine the OzoneManagerProtocol.  And as @cxorm says, that is likely to be a huge patch.

> If that is too much change, we can explore that in a new Jira. Lets
>   get the work going in, as I see it is causing lot of confusion to OM
>   developers

I think that is the right decision. I think there are two separate problems

1. removing the old unused methods, (and fixing the tests that use them).
2. fixing the OzoneManagerProtocol

As of right now, I think they can be handled in separate PR's.  And I think 1 is much higher priority, because as you've said, the code is causing so much confusion.  What do you think, @bharatviswa504 ?

> And also in similar mannner are you also planning to delete
>  KeyManagerImpl delete method also? (I think we can completely remove
>  that method)

Yes, but I think the most efficient way to do it is to finish each class in 'org.apached.hadoop.ozone.om' one at a time.

In other words, finish removing all the methods from OzoneManager.java first.  And fix all the tests that that breaks.  Then all the methods from the KeyManager.java and fix the tests that those break, and so on.  Does that sound ok, @bharatviswa504 ?


>  I ever used to the same throw-Exception manner and I think it doesn't solve the problem fully.

So do you think it is ok to just continue on with the OM for now, and add UnsupportedOperationException's for the rest of the methods, before moving on to the other classes?  Would that cause any problems @cxorm ?

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:2685]: > As of right now, I think they can be handled in separate PR's. And I think 1 is much higher priority, because as you've said, the code is causing so much confusion. What do you think, @bharatviswa504 ?
Yes lets remove old code first which is causing confusion.

> In other words, finish removing all the methods from OzoneManager.java first. And fix all the tests that that breaks. Then all the methods from the KeyManager.java and fix the tests that those break, and so on. Does that sound ok, @bharatviswa504 ?
Yes. I am +1. 

I agree with all other responses also. Lets continue cleanup by class and make progress.

**Comment by GeorgeJahad** [comment]
this one will be completed here:
https://github.com/apache/ozone/pull/2703
